Missing dependencies in generated jvm_import rules leads to runtime exceptions

Aloha!

We're seeing an issue where `jvm_import` rules generated by `maven_install` are missing dependencies in their `deps` fields. This compiles successfully, but leads to runtime exceptions where we cannot load classes from a transitive dependency. At the moment, we’re seeing this with artifacts in `io.grpc`, but this may occur elsewhere too. For reference, we are running **bazel 3.3.0**, **rules_jvm_external 3.3**.

Below is a concrete and reproducible example inspired by a test we saw in [Netflix/concurrency-limits](https://github.com/Netflix/concurrency-limits/blob/94143890c0a2a2339e90f2cf39f326bc79d50454/concurrency-limits-grpc/src/test/java/com/netflix/concurrency/limits/grpc/server/ConcurrencyLimitServerInterceptorTest.java). This is a close representation of actual issues that we are seeing -- in real services that we are migrating to Bazel. (Sorry for the long write-up, but wanted to make sure we had all the details covered!)

---

Say we have a project that directly depends on a standard set of `io.grpc` artifacts, including `io.grpc:grpc-netty-shaded`:

**WORKSPACE**
```
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

RULES_JVM_EXTERNAL_TAG = "3.3"

http_archive(
    name = "rules_jvm_external",
    strip_prefix = "rules_jvm_external-%s" % RULES_JVM_EXTERNAL_TAG,
    url = "https://github.com/bazelbuild/rules_jvm_external/archive/%s.zip" % RULES_JVM_EXTERNAL_TAG,
)

load("@rules_jvm_external//:defs.bzl", "maven_install")

maven_install(
    artifacts = [
        "io.grpc:grpc-api:1.29.0",
        "io.grpc:grpc-netty-shaded:1.29.0",
        "io.grpc:grpc-services:1.29.0",
        "io.grpc:grpc-stub:1.29.0",
    ],
    repositories = [
        "https://repo1.maven.org/maven2",
    ],
)
```

**src/main/java/com/foo/BUILD**
```
load("@rules_java//java:defs.bzl", "java_binary")

java_binary(
    name = "Main",
    srcs = glob(["*.java"]),
    deps = [
        "@maven//:io_grpc_grpc_api",
        "@maven//:io_grpc_grpc_netty_shaded",
        "@maven//:io_grpc_grpc_services",
        "@maven//:io_grpc_grpc_stub",
    ],
)
```

**src/main/java/com/foo/Main.java** (This simply creates a server and calls a single endpoint.)
```
package com.foo;

import io.grpc.CallOptions;
import io.grpc.Channel;
import io.grpc.MethodDescriptor.MethodType;
import io.grpc.MethodDescriptor;
import io.grpc.Server;
import io.grpc.ServerCallHandler;
import io.grpc.ServerServiceDefinition;
import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder;
import io.grpc.netty.shaded.io.grpc.netty.NettyServerBuilder;
import io.grpc.stub.ClientCalls;
import io.grpc.stub.ServerCalls;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;

public class Main {
    public static void main(String[] args) throws IOException {
        final MethodDescriptor<String, String> method = MethodDescriptor.<String, String>newBuilder()
                .setType(MethodType.UNARY)
                .setFullMethodName("service/method")
                .setRequestMarshaller(new FakeMarshaller())
                .setResponseMarshaller(new FakeMarshaller())
                .build();

        final ServerCallHandler<String, String> handler = ServerCalls.asyncUnaryCall((request, observer) -> {
            observer.onNext("response");
            observer.onCompleted();
        });

        final ServerServiceDefinition service = ServerServiceDefinition
                .builder("service")
                .addMethod(method, handler)
                .build();

        final Server server = NettyServerBuilder
                .forPort(9999)
                .addService(service)
                .build();

        server.start();

        final Channel channel = NettyChannelBuilder
                .forAddress("localhost", server.getPort())
                .usePlaintext()
                .build();

        ClientCalls.blockingUnaryCall(channel, method, CallOptions.DEFAULT, "request");
    }
}

class FakeMarshaller implements MethodDescriptor.Marshaller<String> {
    public InputStream stream(String var1) {
        return new ByteArrayInputStream(new byte[0]);
    }

    public String parse(InputStream var1) {
        return "";
    }
}
```

At runtime (`bazel run //...`), we will see a `NoClassDefFoundError` exception related to `io.perfmark`:
```
Exception in thread "main" java.lang.NoClassDefFoundError: io/perfmark/PerfMark
	at io.grpc.internal.ClientCallImpl.<init>(ClientCallImpl.java:109)
	at io.grpc.internal.ManagedChannelImpl$RealChannel.newCall(ManagedChannelImpl.java:891)
	at io.grpc.internal.ServiceConfigInterceptor.interceptCall(ServiceConfigInterceptor.java:137)
	at io.grpc.ClientInterceptors$InterceptorChannel.newCall(ClientInterceptors.java:156)
	at io.grpc.internal.ManagedChannelImpl.newCall(ManagedChannelImpl.java:856)
	at io.grpc.internal.ForwardingManagedChannel.newCall(ForwardingManagedChannel.java:63)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:128)
	at com.foo.Main.main(Main.java:50)
Caused by: java.lang.ClassNotFoundException: io.perfmark.PerfMark
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 8 more
```

**Why does this happen?**

First, please note that a dependency path exists from `io.grpc:grpc-netty-shaded` -> `io.grpc:grpc-core` -> `io.perfmark:perfmark-api`.

Next, if we look into `./bazel-foo/external/maven/BUILD`, we will see:

```
jvm_import(
	name = "io_grpc_grpc_netty_shaded",
	jars = ["v1/https/repo1.maven.org/maven2/io/grpc/grpc-netty-shaded/1.29.0/grpc-netty-shaded-1.29.0.jar"],
	deps = [
		":io_grpc_grpc_core",
	],
	tags = ["maven_coordinates=io.grpc:grpc-netty-shaded:1.29.0"],
)

# ...

jvm_import(
	name = "io_grpc_grpc_core",
	jars = ["v1/https/repo1.maven.org/maven2/io/grpc/grpc-core/1.29.0/grpc-core-1.29.0.jar"],
	deps = [
	],
	tags = ["maven_coordinates=io.grpc:grpc-core:1.29.0"],
)

# ...

jvm_import(
	name = "io_perfmark_perfmark_api",
	jars = ["v1/https/repo1.maven.org/maven2/io/perfmark/perfmark-api/0.19.0/perfmark-api-0.19.0.jar"],
	deps = [
		":com_google_code_findbugs_jsr305",
	],
	tags = ["maven_coordinates=io.perfmark:perfmark-api:0.19.0"],
)
```

The `deps` field for `io_grpc_grpc_core` is empty. In fact, nothing in the entire `BUILD` file refers to `io_perfmark_perfmark_api`. This is problematic later on when we run our program, because classes from `io.perfmark:perfmark-api` will never be placed on the runtime classpath, leading to the `NoClassDefFoundError` exception.

Let's also look at the output from Coursier in `./bazel-foo/external/maven/dep-tree.json`:

```
{
  "conflict_resolution": {},
  "dependencies": [

    ...

    {
      "coord": "io.grpc:grpc-netty-shaded:1.29.0",
      "file": "v1/https/repo1.maven.org/maven2/io/grpc/grpc-netty-shaded/1.29.0/grpc-netty-shaded-1.29.0.jar",
      "directDependencies": [
        "io.grpc:grpc-core:1.29.0"
      ],
      "dependencies": [
        "io.grpc:grpc-core:1.29.0"
      ]
    },

    ...

    {
      "coord": "io.grpc:grpc-core:1.29.0",
      "file": "v1/https/repo1.maven.org/maven2/io/grpc/grpc-core/1.29.0/grpc-core-1.29.0.jar",
      "directDependencies": [],
      "dependencies": []
    },

    ...

    {
      "coord": "io.perfmark:perfmark-api:0.19.0",
      "file": "v1/https/repo1.maven.org/maven2/io/perfmark/perfmark-api/0.19.0/perfmark-api-0.19.0.jar",
      "directDependencies": [
        "com.google.code.findbugs:jsr305:3.0.2"
      ],
      "dependencies": [
        "com.google.code.findbugs:jsr305:3.0.2"
      ],
      "exclusions": [
        "com.google.errorprone:error_prone_annotations"
      ]
    },

    ...

  ],
  "version": "0.1.0"
}
```

There are missing dependencies here too. Given how we derive the `jvm_import` rules from the Coursier output, this explains what we saw in the `BUILD` file. We can also sanity check by running `coursier fetch io.grpc:grpc-netty-shaded:1.29.0 --json-output-file tree.json` -- we will see the same behavior.

Therefore, this looks like a Coursier bug that is causing issues when using `rules_jvm_external`. There is already an issue with similar symptoms (but possibly different cause?) reported over there @ https://github.com/coursier/coursier/issues/1708, but we wanted to make sure this is surfaced here too, in case there is the possibility of an intermediate fix that can be placed into `rules_jvm_external` itself.

Thank you for reading!

--- a/WORKSPACE
+++ b/WORKSPACE
@@ -383,6 +383,17 @@ maven_install(
     fetch_sources = True,
 )
 
+# https://github.com/bazelbuild/rules_jvm_external/issues/433
+maven_install(
+    name = "version_interval_testing",
+    artifacts = [
+        "io.grpc:grpc-netty-shaded:1.29.0",
+    ],
+    repositories = [
+        "https://repo1.maven.org/maven2",
+    ],
+)
+
 load(
     "@json_artifacts_testing//:defs.bzl",
     _json_artifacts_testing_install = "pinned_maven_install",

--- a/tests/integration/BUILD
+++ b/tests/integration/BUILD
@@ -37,6 +37,25 @@ diff_test(
     file2 = ":testonly-deps",
 )
 
+# https://github.com/coursier/coursier/issues/1792
+# https://github.com/bazelbuild/rules_jvm_external/issues/433
+genquery(
+    name = "version_interval_deps",
+    expression = "deps(@version_interval_testing//:io_grpc_grpc_netty_shaded, 2)",
+    opts = ["--nohost_deps", "--noimplicit_deps"],
+    scope = ["@version_interval_testing//:io_grpc_grpc_netty_shaded"],
+    testonly = True,
+)
+
+diff_test(
+    name = "version_interval_deps_test",
+    file1 = select({
+        "@bazel_tools//src/conditions:windows": "version-interval-deps.golden.dos",
+        "//conditions:default": "version-interval-deps.golden.unix",
+    }),
+    file2 = ":version_interval_deps",
+)
+
 java_import(
     name = "guava_import",
     jars = [

--- a/tests/integration/version-interval-deps.golden.dos
+++ b/tests/integration/version-interval-deps.golden.dos
@@ -1,0 +1,9 @@
+@version_interval_testing//:io_grpc_grpc_netty_shaded
+@version_interval_testing//:v1/https/repo1.maven.org/maven2/io/grpc/grpc-netty-shaded/1.29.0/grpc-netty-shaded-1.29.0.jar
+@version_interval_testing//:io_grpc_grpc_core
+@version_interval_testing//:v1/https/repo1.maven.org/maven2/io/grpc/grpc-core/1.29.0/grpc-core-1.29.0.jar
+@version_interval_testing//:io_perfmark_perfmark_api
+@version_interval_testing//:io_grpc_grpc_api
+@version_interval_testing//:com_google_errorprone_error_prone_annotations
+@version_interval_testing//:com_google_code_gson_gson
+@version_interval_testing//:com_google_android_annotations

--- a/tests/integration/version-interval-deps.golden.unix
+++ b/tests/integration/version-interval-deps.golden.unix
@@ -1,0 +1,9 @@
+@version_interval_testing//:io_grpc_grpc_netty_shaded
+@version_interval_testing//:v1/https/repo1.maven.org/maven2/io/grpc/grpc-netty-shaded/1.29.0/grpc-netty-shaded-1.29.0.jar
+@version_interval_testing//:io_grpc_grpc_core
+@version_interval_testing//:v1/https/repo1.maven.org/maven2/io/grpc/grpc-core/1.29.0/grpc-core-1.29.0.jar
+@version_interval_testing//:io_perfmark_perfmark_api
+@version_interval_testing//:io_grpc_grpc_api
+@version_interval_testing//:com_google_errorprone_error_prone_annotations
+@version_interval_testing//:com_google_code_gson_gson
+@version_interval_testing//:com_google_android_annotations

